Stratis.Bitcoin/Consensus/ConsensusFeature.cs

```diff
@@ -0,0 +1,187 @@
+                    {
+                        Logs.FullNode.LogError("Block rejected: " + block.Error.Message);
+
+                        //Pull again
+                        this.consensusLoop.Puller.SetLocation(this.consensusLoop.Tip);
+
+                        if (block.Error == ConsensusErrors.BadWitnessNonceSize)
+                        {
+                            Logs.FullNode.LogInformation("You probably need witness information, activating witness requirement for peers.");
+                            this.connectionManager.AddDiscoveredNodesRequirement(NodeServices.NODE_WITNESS);
+                            this.consensusLoop.Puller.RequestOptions(TransactionOptions.Witness);
+                            continue;
+                        }
+
+                        //Set the PoW chain back to ConsensusLoop.Tip
+                        this.chain.SetTip(this.consensusLoop.Tip);
+                        //Since ChainBehavior check PoW, MarkBlockInvalid can't be spammed
+                        Logs.FullNode.LogError("Marking block as invalid");
+                        this.chainState.MarkBlockInvalid(block.ChainedBlock.HashBlock);
+                    }
+
+                    if (!reorg && block.Error == null)
+                    {
+                        this.chainState.HighestValidatedPoW = this.consensusLoop.Tip;
+                        if (this.chain.Tip.HashBlock == block.ChainedBlock?.HashBlock)
+                        {
+                            var unused = cache.FlushAsync();
+                        }
+
+                        this.signals.Blocks.Broadcast(block.Block);
+                    }
+
+                    // TODO: replace this with a signalling object
+                    if (stats.CanLog)
+                        stats.Log();
+                }
+            }
+            catch (Exception ex) 
+            {
+                if (ex is OperationCanceledException)
+                {
+                    if (this.globalCancellation.Cancellation.IsCancellationRequested)
+                        return;
+                }
+
+                // TODO Need to revisit unhandled exceptions in a way that any process can signal an exception has been
+                // thrown so that the node and all the disposables can stop gracefully.
+                Logs.Consensus.LogCritical(new EventId(0), ex, "Consensus loop unhandled exception (Tip:" + this.consensusLoop.Tip?.Height + ")");
+                throw;
+            }
+        }        
+    }
+
+    public static class ConsensusFeatureExtension
+    {
+        public static IFullNodeBuilder UseConsensus(this IFullNodeBuilder fullNodeBuilder)
+        {
+            
+            fullNodeBuilder.ConfigureFeature(features =>
+            {
+                features
+                .AddFeature<ConsensusFeature>()
+                .FeatureServices(services =>
+                {
+                    var dataFolder = new DataFolder(fullNodeBuilder.NodeSettings);
```

Comment:
This should be injected but I guess that's fine for now @Neurosploit is fixing this in his next commit.